### PR TITLE
Add a notebook example to reach a quick baseline of ~94% accuracy on CIFAR

### PR DESCRIPTION
**Pull request by hemildesai**
…CIFAR10 using Resnet in Lightning

## What does this PR do?

This PR adds a notebook to reach a baseline of ~94% accuracy using Resnet on CIFAR10 using pytorch-lightning. Related Conversation - [Slack](https://pytorch-lightning.slack.com/archives/CQXV8BRH9/p1605710339026200)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements) (Approved on Slack)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃
Yesss

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Great notebook !

File: notebooks/README.md [hunk 1]
```diff
@@ -4,10 +4,11 @@
...
+| **BERT**                 | Fine-tune HuggingFace Transformers models on the GLUE Benchmark                      | [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/04-transformers-text-classification.ipynb) |
+| **Trainer Flags**        | Overview of the available Lightning `Trainer` flags                                  |      [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/05-trainer-flags-overview.ipynb)      |
+| **94% Baseline CIFAR10** | Establish a quick baseline of ~94% accuracy on CIFAR10 using Resnet in Lightning     |         [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/06-cifar10-baseline.ipynb)         |
```
Comment (rohitgr7) [inline notebooks/README.md:14]:
```suggestion
| **94% Baseline CIFAR10** | Establish a quick baseline of ~94% accuracy on CIFAR10 with Resnet using Lightning     |         [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/PytorchLightning/pytorch-lightning/blob/master/notebooks/06-cifar10-baseline.ipynb)         |
```

**Review by rohitgr7 — COMMENTED** [review]
I'd suggest `restart and clear output`, just like other notebooks maybe.

**Comment by rohitgr7** [comment]
not sure how to edit a notebook here so putting some suggestions here:
```
cifar10_dm = pl_bolts.datamodules.CIFAR10DataModule(batch_size=batch_size)
cifar10_dm.train_transforms = train_transforms
cifar10_dm.val_transforms = test_transforms
cifar10_dm.test_transforms = test_transforms
```
you can pass them as:
```python
cifar10_dm = pl_bolts.datamodules.CIFAR10DataModule(
    batch_size=batch_size,
    train_transforms=train_transforms,
    val_transforms=val_transforms,
    test_transforms=test_transforms
)
```
---

remove this. let trainer call it:
```python
cifar10_dm.prepare_data()
```
---

add 4 space indentation.

---
use LearningRateMonitor
```python
self.log('Learning Rate', self.optimizers().param_groups[0]['lr'])
```
---
update
```python
torch.optim.SGD(self.model.parameters(), lr=self.hparams.lr, momentum=0.9, weight_decay=5e-4)
```
to
```python
torch.optim.SGD(self.parameters(), lr=self.hparams.lr, momentum=0.9, weight_decay=5e-4)
```

---
```
update_bn(dl, self.swa_model.to(self.device), device=self.device)
```

**Comment by Borda** [comment]
> not sure how to edit a notebook here so putting some suggestions here:

check the link from reviewNB
https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/

**Comment by Borda** [comment]
@hemildesai mind integrate suggestion from @rohitgr7 then shall be fine to go...

**Comment by rohitgr7** [comment]
> check the link from reviewNB
https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/

nice! Didn't know about this.

File: notebooks/06-cifar10-baseline.ipynb [hunk 2]
```diff
@@ -0,0 +1,384 @@
+{
```
Comment (Borda) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p>lets remove the note about unresolved issue...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Comment by hemildesai** [comment]
> update_bn(dl, self.swa_model.to(self.device), device=self.device)

@rohitgr7 I tried this but for some reason `self.device` was returning `cpu` since I need to call it separately and outside the `trainer.fit` method

**Comment by rohitgr7** [comment]
> @rohitgr7 I tried this but for some reason self.device was returning cpu since I need to call it separately and outside the trainer.fit method

will try to find an alternative for this.

File: notebooks/06-cifar10-baseline.ipynb [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
```
Comment (rohitgr7) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p>I'd suggest using <code>-qU</code>  instead of <code>--quiet</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (rohitgr7) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p><code>self.log('train_loss', loss)</code></p><p>just to match it with other keys.</p><p>---</p><p> </p><p>mind clean configure_optimizer a bit? since notebook render on github is not so good.</p><pre class="ql-syntax" spellcheck="false"># add in imports
from torch.optim.lr_scheduler import OneCycleLR

optimizer = ...
steps_per_epoch = 

lr_dict = {
    'scheduler': OneCycleLR(optimizer, 0.1, epoch=..., steps_per_epoch=...),
    'interval': 'step',
}

return {
    'optimizer': optimizer,
    'lr_scheduler': lr_dict,
}
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (rohitgr7) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p>can you try <code>update_bn(...)</code> in <code>on_train_end</code> hook instead of calling it explicitly after <code>trainer.fit</code>?</p><p><code>update_bn(self.datamodule.train_dataloader(), self.swa_model, device=self.device)</code></p><p> </p><p>also what's the significance of using SWA before testing?</p><p> </p><p>Also I'd suggest creating a base model and overriding the methods that are not same in both, just to avoid duplicate code.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (rohitgr7) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p>should it be <code>lightining_logs/version_0</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (rohitgr7) [inline notebooks/06-cifar10-baseline.ipynb:1]: <p>I think it's <code>Bolts</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/PyTorchLightning/pytorch-lightning/pull/4818/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (hemildesai) [reply notebooks/06-cifar10-baseline.ipynb:1]: <p>Yes I will try <code>update_bn</code> in <code>on_train_end</code>.</p><p> </p><p>SWA is a method to improve generalization when using SGD. There's more details at https://pytorch.org/blog/stochastic-weight-averaging-in-pytorch/ and also in the original paper at https://arxiv.org/abs/1803.05407</p>

Comment (hemildesai) [reply notebooks/06-cifar10-baseline.ipynb:1]: <p><code>lightning_logs</code> is the default directory for tensorboard logs so we need to provide it as the path in order to see all versions and runs.</p>

Comment (hemildesai) [reply notebooks/06-cifar10-baseline.ipynb:1]: <p>This was taken from https://github.com/PyTorchLightning/pytorch-lightning/pull/4555</p>

Comment (hemildesai) [reply notebooks/06-cifar10-baseline.ipynb:1]: Regarding inheritance, the SWAResnet inherits from LitResnet and overrides the methods that need change.

Comment (rohitgr7) [reply notebooks/06-cifar10-baseline.ipynb:1]: Although it's correct but I think it can be done in a better way using `load_from_checkpoint` and keeping Model init same for both. Can I make a commit with some changes here by the weekend, if that's okay with you??

Comment (hemildesai) [reply notebooks/06-cifar10-baseline.ipynb:1]: Yeah sure! Go for it!

Comment (rohitgr7) [reply notebooks/06-cifar10-baseline.ipynb:1]: > the SWAResnet inherits from LitResnet and overrides the methods that need change.

it was all good.

made some changes. BTW thanks for the SWA link.

**Review by rohitgr7 — APPROVED** [review]

**Review by rohitgr7 — CHANGES_REQUESTED** [review]
Pls wait. Don't merge. Just realised, missed one thing.

**Comment by Borda** [comment]
> Pls wait. Don't merge. Just realised, missed one thing.

Use the wip in the PR name for this case, it automatically block checks and prevent merging =)

**Review by SuoerMan — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]
lgtm
